stdlib/public/runtime/Casting.cpp

```diff
@@ -2755,6 +2755,50 @@ swift::swift_conformsToProtocol(const Metadata *type,
   goto recur;
 }
 
+/// Return the type metadata for a given type name.
+///
+/// This is a temporary implementation to support some Foundation use cases;
+/// it will only lookup a type that directly implements a protocol as only
+/// those types will have conformance table entries.
+static const Metadata *
+_metadataForName(const std::string &typeName) {
+  auto &C = Conformances.get();
+  const Metadata *foundMetadata = nullptr;
+
+  installCallbacksToInspectDylib();
+
+  pthread_mutex_lock(&C.SectionsToScanLock);
+
+  unsigned sectionIdx = 0;
+  unsigned endSectionIdx = C.SectionsToScan.size();
+
+  // TODO: implement typeName->metadata cache
+  // TODO: support looking up types for which there is no conformance entry
```

Comment:
You should also note that this doesn't handle generic or non-nominal types yet (which is fine, for a start).